worker/docs/buildworker.1
.\" This file is part of Buildbot.  Buildbot is free software: you can
.\" redistribute it and/or modify it under the terms of the GNU General Public
.\" License as published by the Free Software Foundation, version 2.
.\" 
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
.\" FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
.\" details.
.\" 
.\" You should have received a copy of the GNU General Public License along with
.\" this program; if not, write to the Free Software Foundation, Inc., 51
.\" Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\" 
.\" Copyright Buildbot Team Members

.TH BUILDWORKER "1" "August 2010" "Buildbot" "User Commands"
.SH NAME
buildworker \- a tool for managing buildbot worker instances
.SH SYNOPSIS
.PP
.B buildworker
[
.BR "global options"
]
.I command
[
.BR "command options"
]
.PP
.B buildworker
create-worker
[
.BR \-q | \-\-quiet
]
[
.BR \-f | \-\-force
]
[
.BR \-r | \-\-relocatable
]
[
.BR \-n | \-\-no-logrotate
]
[
.BR \-k | \-\-keepalive
.I TIME
]
[
.BR --usepty
{0|1}
]
[
.BR \-\-umask
.I UMASK
]
[
.BR \-s | \-\-log-size
.I SIZE
]
[
.BR \-l | \-\-log-count
.I COUNT
]
[
.BR \-\-verbose
]
.I PATH
.I MASTER
.I USERNAME
.I PASSWORD
.PP
.B buildworker
[
.BR \-\-verbose
]
{
.BR start | stop | restart
}
[
.I PATH
]
.PP
.B buildworker
[
.BR \-\-verbose
]
{
.BR \-\-help | \-\-version
}
.PP
.B buildworker
.I command
.BR \-h | \-\-help
.SH DESCRIPTION
.\" Putting a newline after each sentence can generate better output.
The `buildworker' command-line tool can be used to start or stop a
buildworker or create a new buildworker instance.
.SH OPTIONS
.SS Commands
.TP
.BR create-worker
Create and populate a directory for a new buildworker
.TP
.BR start
Start a buildworker
.TP
.BR stop
Stop a buildworker
.TP
.BR restart
Restart a buildworker
.SS Global options
.TP
.BR \-h | \-\-help
Print the list of available commands and global options.
All subsequent commands are ignored.
.TP
.BR --version
Print twistd and buildworker version.
All subsequent commands are ignored.
.TP
.BR --verbose
Verbose output.
.SS create-worker command options
.TP
.BR \-f | \-\-force
Re-use an existing directory.
.TP
.BR \-h | \-\-help
Show help for current command and exit.
All subsequent commands are ignored.
.TP
.BR \-k | \-\-keepalive
Send keepalive requests to buildmaster every
.I TIME
seconds.
Default value is 600 seconds.
.TP
.BR \-l | \-\-log-count
Limit the number of kept old twisted log files to
.IR COUNT .
All files are kept by default.
.TP
.BR \-q | \-\-quiet
Do not emit the commands being run.
.TP
.BR \-r | \-\-relocatable
Create a relocatable buildbot.tac.
.TP
.BR \-n | \-\-no-logrotate
Do not permit buildworker rotate logs by itself.
.TP
.BR \-s | \-\-log-size
Set size at which twisted lof file is rotated to
.I SIZE
bytes.
Default value is 1000000 bytes.
.TP
.BR \-\-umask
Set umask for files created by buildworker.
Default value is 077 which means only owner can access the files.
See
.BR umask (2)
for more details.
.TP
.BR \-\-usepty
Set wether child processes should be run in a pty (0 means do not run in a
pty).
Default value is 0.
.TP
.I PATH
Path to buildworker base directory.
.TP
.I MASTER
Set the host and port of buildbot master to attach to in form
.IR HOST:PORT .
This should be provided by buildmaster administrator.
.TP
.I USERNAME
Buildworker name to connect with.
This should be provided by buildmaster administrator.
.TP
.I PASSWORD
Buildworker password to connect with.
This should be provided by buildmaster administrator.
.SH "SEE ALSO"
.BR buildbot (1),
.BR umask (2),
.PP
The complete documentation is available in texinfo format. To use it, run
.BR "info buildbot" .



NEWS
Major User visible changes in Buildbot.             -*- outline -*-
  (see ChangeLog for a detailed list of changes)

* Release 0.8.0p1 (October 1, 2010)

** Fix for XSS vulnerabilities discovered by Amber Yust

* Release 0.8.0 (May 25, 2010)

** (NOTE!) Scheduler requires keyword arguments

If you are creating your Scheduler like this:

  Scheduler("mysched", "mybranch", 0, ["foo", "bar"])

then it's time to change that to specify each of the arguments with a keyword:

  Scheduler(name="mysched", branch="mybranch",
        treeStableTimer=0, builderNames=["foo", "bar"])

** Database Backend

Scheduler, change, and build request information is now stored in a database -
by default, in SQLite, although MySQL is also supported.  With this change,
scheduled builds will persist over buildmaster restarts, as will
interrelationships between schedulers (e.g., Triggerable and Dependent).

Upgrading to the new database backend is easy, although it brings additional
requirements on the buildmaster.  See the Buildbot documentation for more
information.

** Visual Studio / VC++ Compile Steps

** New Change/SourceStamp attributes 'project' and 'repository'

These attributes can be used to further refine matching by schedulers.
Repository completes the SourceStamp: the tuple of (repository, branch,
revision) completely specifies a source code tree.  Likewise, the project
attribute can be used to support building several distinct projects within one
buildmaster, replacing the use of category for this purpose.  Matching can be
done using regular expressions, so it's even possible to support nested
projects!

** ShellCommands expand environment variables

If you pass to a shell command an environment variable like this:

ShellCommand(..., env={"FOO": "${BAR}"})

then, on the slave side the variable FOO will have the same value as
the alread existing BAR variable on the slave. This is mostly used to
expand variable like this: "PATH": "/my/directory:${PATH}" where PATH
will have "/my/directory" prepended to it.

** Builders can setup properties

There is a new parameter to the builders to setup properties on a
per-builder basis.

** New /json web status

This view has lots of useful information perfectly formed for serving as input
to JavaScript status displays.  See /json/help for details.

** Jinja

All web status is now generated using the Jinja templating engine, which gives
buildbot a much more attractive and maintainable appearance.  Buildbot's output
is also now XHTML-compliant!

** Authorization Framework

The web-based status displays now provide fine-grained control over who can
do what - force builds, stop builds, cancel builds, etc.  See the manual for
configuration details.

** Mercurial uses full revisions

Mercurial now sets got_revision to the full 40-character revision id instead of
the short IDs.

** Cleanup, Bug Fixes, and Test Fixes

Thanks to help from a number of devoted contributors, this version of Buildbot
has seen a lot of house-cleaning, and even passes all of its own unit tests!  

** Removals

*** Removed buildbot.status.html.Waterfall (deprecated in 0.7.6)

Note that this does not remove the waterfall -- just an old version of it which
did not include the rest of the WebStatus pages.

*** BuildmasterConfig no longer accepts 'bots' and 'sources' as keys
(deprecated in 0.7.6). Use 'slaves' and 'change_source' instead.

* Release 0.7.12 (January 21, 2010)

** New 'console' display

This is a new web status view combining the best of the (t)grid and waterfall
views.

** New 'extended' stylesheet

Buildbot has a new, much nicer stylesheet available.  Copy the file
buildbot/status/web/extended.css over your existing public_html/buildbot.css to
se it.

** Builders can be configured with an object

Instead of a list of dictionaries, builders can now specified using a
BuilderConfig object in the configuration file.  This will allow for better
argument checking and default values, and also makes it easier for users to
create subclasses to handle site-specific builder details.  The old,
dictionary-based method of configuration is still supported.

** Check for common mis-configuration in addStep

When adding a new step to a factory, either of these are acceptable:

  f.addStep(ShellCommand(command="echo hello, world", description="say hi"))
  f.addStep(ShellCommand, command="echo hello, world", description="say hi")

but trying to mix these syntaxes is a common misconfiguration:

  f.addStep(ShellCommand(command="echo hello, world"), description="say hi")

in which case the description argument was silently ignored.  This is now an
error.

** Support for log compression

Log files can be compressed on the master side using either gzip or bzip2.

** Builder.ping no longer accepts timeout argument (bug #664).

The implementation was not robust enough and could cause the master to
unexpectedly disconnect the slave.

** MailNotifier's customMesg replaced by messageFormatter

The customMesg mechanism had the unfortunate side effect of loading all data
for a build into memory simultaneously, which for some builds could cause
memory exhaustion.

** Suppression of selected compiler warnings

The WarningCountingShellCommand class has been extended with the ability to
upload from the slave a file contain warnings to be ignored. See the
documentation of the suppressionFile argument to the Compile build step.

** New buildstep `MTR'

A new class buildbot.process.mtrlogobserver.MTR was added. This buildstep is
used to run test suites using mysql-test-run. It parses the stdio output for
test failures and summarises them on the waterfall page. It also makes server
error logs available for debugging failures, and optionally inserts
information about test runs and test failures into an external database.

** Python API Docs

The docstrings for buildbot are now available in a web-friendly format:
  http://djmitche.github.com/buildbot/docs/latest/reference

** Many, many bugfixes


worker/contrib/init-scripts/buildworker.init.sh
#!/bin/bash

### Maintain compatibility with chkconfig
# chkconfig: 2345 83 17
# description: buildworker

### BEGIN INIT INFO
# Provides:          buildworker
# Required-Start:    $remote_fs
# Required-Stop:     $remote_fs
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Buildbot worker init script
# Description:       This file allows running buildbot worker instances at
#                    startup
### END INIT INFO

PATH=/sbin:/bin:/usr/sbin:/usr/bin
WORKER_RUNNER=/usr/bin/buildworker


# Source buildworker configuration
[[ -r /etc/default/buildworker ]] && . /etc/default/buildworker
#[[ -r /etc/sysconfig/buildworker ]] && . /etc/sysconfig/buildworker

# Or define/override the configuration here
#WORKER_ENABLED[1]=0                    # 0-enabled, other-disabled
#WORKER_NAME[1]="buildworker #1"         # short name printed on start/stop
#WORKER_USER[1]="buildbot"              # user to run worker as
#WORKER_BASEDIR[1]=""                   # basedir to worker (absolute path)
#WORKER_OPTIONS[1]=""                   # buildbot options
#WORKER_PREFIXCMD[1]=""                 # prefix command, i.e. nice, linux32, dchroot


# Get some LSB-like functions
if [ -r /lib/lsb/init-functions ]; then
    . /lib/lsb/init-functions
else
    function log_success_msg() {
        echo "$@"
    }
    function log_failure_msg() {
        echo "$@"
    }
    function log_warning_msg() {
        echo "$@"
    }
fi


# Some systems don't have seq (e.g. Solaris)
if type seq >/dev/null 2>&1; then
    :
else
    function seq() {
        for ((i=1; i<=$1; i+=1)); do
            echo $i
        done
    }
fi


if [[ ! -x ${WORKER_RUNNER} ]]; then
    log_failure_msg "does not exist or not an executable file: ${WORKER_RUNNER}"
    exit 1
fi

function is_enabled() {
    ANSWER=`echo $1|tr "[:upper:]" "[:lower:]"`
    [[ "$ANSWER" == "yes" ]] || [[ "$ANSWER" == "true" ]] || [[ "$ANSWER" ==  "1" ]]
    return $?
}

function is_disabled() {
    ANSWER=`echo $1|tr "[:upper:]" "[:lower:]"`
    [[ "$ANSWER" == "no" ]] || [[ "$ANSWER" == "false" ]] || [[ "$ANSWER" ==  "0" ]]
    return $?
}


function worker_config_valid() {
    # Function validates buildworker instance startup variables based on array
    # index
    local errors=0
    local index=$1

    if ! is_enabled "${WORKER_ENABLED[$index]}" && ! is_disabled "${WORKER_ENABLED[$index]}" ; then
        log_warning_msg "buildworker #${index}: invalid enabled status"
        errors=$(($errors+1))
    fi

    if [[ -z ${WORKER_NAME[$index]} ]]; then
        log_failure_msg "buildworker #${index}: no name"
        errors=$(($errors+1))
    fi

    if [[ -z ${WORKER_USER[$index]} ]]; then
        log_failure_msg "buildworker #${index}: no run user specified"
        errors=$( ($errors+1) )
    elif ! getent passwd ${WORKER_USER[$index]} >/dev/null; then
        log_failure_msg "buildworker #${index}: unknown user ${WORKER_USER[$index]}"
        errors=$(($errors+1))
    fi

    if [[ ! -d "${WORKER_BASEDIR[$index]}" ]]; then
        log_failure_msg "buildworker ${index}: basedir does not exist ${WORKER_BASEDIR[$index]}"
        errors=$(($errors+1))
    fi

    return $errors
}

function check_config() {
    itemcount="${#WORKER_ENABLED[@]}
               ${#WORKER_NAME[@]}
               ${#WORKER_USER[@]}
               ${#WORKER_BASEDIR[@]}
               ${#WORKER_OPTIONS[@]}
               ${#WORKER_PREFIXCMD[@]}"

    if [[ $(echo "$itemcount" | tr -d ' ' | sort -u | wc -l) -ne 1 ]]; then
        log_failure_msg "WORKER_* arrays must have an equal number of elements!"
        return 1
    fi

    errors=0
    for i in $( seq ${#WORKER_ENABLED[@]} ); do
        if is_disabled "${WORKER_ENABLED[$i]}" ; then
            log_warning_msg "buildworker #${i}: disabled"
            continue
        fi
        worker_config_valid $i
        errors=$(($errors+$?))
    done

    [[ $errors == 0 ]]; return $?
}

check_config || exit $?

function iscallable () { type $1 2>/dev/null | grep -q 'shell function'; }

function worker_op () {
    op=$1 ; mi=$2

    if [ `uname` = SunOS ]; then
        suopt=""
    else
        suopt="-s /bin/sh"
    fi
    ${WORKER_PREFIXCMD[$mi]} \
    su $suopt - ${WORKER_USER[$mi]} \
    -c "$WORKER_RUNNER $op ${WORKER_OPTIONS[$mi]} ${WORKER_BASEDIR[$mi]} > /dev/null"
    return $?
}

function do_op () {
    errors=0
    for i in $( seq ${#WORKER_ENABLED[@]} ); do
        if [ -n "$4" ] && [ "$4" != "${WORKER_NAME[$i]}" ] ; then
            continue
        elif is_disabled "${WORKER_ENABLED[$i]}" && [ -z "$4" ] ; then
            continue
        fi

        # Some rhels don't come with all the lsb goodies
        if iscallable log_daemon_msg; then
	        log_daemon_msg "$3 \"${WORKER_NAME[$i]}\""
            if eval $1 $2 $i; then
                log_end_msg 0
            else
                log_end_msg 1
                errors=$(($errors+1))
            fi
        else
            if eval $1 $2 $i; then
                log_success_msg "$3 \"${WORKER_NAME[$i]}\""
            else
                log_failure_msg "$3 \"${WORKER_NAME[$i]}\""
                errors=$(($errors+1))
            fi
        fi
    done
    return $errors
}

case "$1" in
    start)
        do_op "worker_op" "start" "Starting buildworker" "$2"
        exit $?
        ;;
    stop)
        do_op "worker_op" "stop" "Stopping buildworker" "$2"
        exit $?
        ;;
    reload)
        do_op "worker_op" "reload" "Reloading buildworker" "$2"
        exit $?
        ;;
    restart|force-reload)
        do_op "worker_op" "restart" "Restarting buildworker" "$2"
        exit $?
        ;;
    *)
        echo "Usage: $0 {start|stop|restart|reload|force-reload}"
        exit 1
        ;;
esac

exit 0
